### PR TITLE
Added titus-network mode support & v6, v4, and TSA awareness

### PR DESCRIPTION
**Pull request by solarkennedy**
This new network-mode is designed to be a "source of truth"
for what kind of networking to setup on a container.

Specifically I want some way to know for sure if TSA and VPCservice
are supposed to setup a ipv4 transition mechanism.

OR, in the future case where we know we want ipv6 only, vpcservice
needs to know that and forgo certain things.

Likewise, titus-inject-metadata will probably have to behave
differently. And also I even expect user applications to take advantage
of the `TITUS_NETWORK_MODE=IPV6_ONLY` style hints, so that they
don't have to scrape ifconfig or other weirdness, and can adjust
their behavior appropriately.

This just plumbs the setting through, no additional code is in place
yet to actually act on it.

### CODE REVIEW

File: cmd/titus-vpc-tool/setup_container.go [hunk 1]
```diff
@@ -33,6 +33,7 @@ func setupContainercommand(ctx context.Context, v *pkgviper.Viper, iipGetter ins
 	}
 
 	cmd.Flags().Int("netns", 3, "The File Descriptor # of the network namespace to setup")
+	cmd.Flags().Int("network-mode", 0, "Type of networking to setup, uses the enum")
```
Comment (sargun) [inline cmd/titus-vpc-tool/setup_container.go:36]: :-/. Why not make this a string?

File: executor/runtime/types/container.go [hunk 2]
```diff
@@ -1301,3 +1312,19 @@ func isEFSID(FsID string) (bool, error) {
 	}
 	return matched, nil
 }
+
+func getHumanFriendlyNetworkMode(mode int) string {
```
Comment (sargun) [inline executor/runtime/types/container.go:1316]: iirc the protobuf definitions have a helper that does this translation.

File: executor/runtime/types/pod.go [hunk 3]
```diff
@@ -394,6 +395,16 @@ func (c *PodContainer) MetatronCreds() *titus.ContainerInfo_MetatronCreds {
 	return c.titusInfo.GetMetatronCreds()
 }
 
+func (c *PodContainer) NetworkMode() int {
+	// TODO: move to pod-common when stable
+	modeStr := c.pod.Annotations["titusParameter.agent.networkMode"]
+	mode, err := strconv.ParseInt(modeStr, 10, 32)
```
Comment (sargun) [inline executor/runtime/types/pod.go:401]: IIRC there is a https://golang.org/pkg/strconv/#Itoa function (Itoa)

**Review by sargun — CHANGES_REQUESTED** [review]

File: cmd/titus-vpc-tool/setup_container.go (hunk 1 above)
Comment (solarkennedy) [reply cmd/titus-vpc-tool/setup_container.go:36]: I can do that.

Ooo yea looks like in the .pb.go it has all the mappers I want to make this sane. Great.

File: executor/runtime/types/container.go [hunk 4]
```diff
@@ -1318,3 +1330,25 @@ func isEFSID(FsID string) (bool, error) {
 	}
 	return matched, nil
 }
+
+// GetHumanFriendlyNetworkMode uses the incoming network mode string
+// and mutates it a bit to be a environment-variable safe string.
+// In the unknown mode, however, we return an empty string for
+// the caller to *not* set the variable
+func GetHumanFriendlyNetworkMode(mode string) string {
```
Comment (sargun) [inline executor/runtime/types/container.go:1338]: In the proto object, there are two maps that have this info, I think?
```

var NetworkConfiguration_NetworkMode_name = map[int32]string{
        0: "UnknownNetworkMode",
        1: "Ipv4Only",
        2: "Ipv6AndIpv4",
        3: "Ipv6AndIpv4Fallback",
        4: "Ipv6Only",
}

var NetworkConfiguration_NetworkMode_value = map[string]int32{
        "UnknownNetworkMode":  0,
        "Ipv4Only":            1,
        "Ipv6AndIpv4":         2,
        "Ipv6AndIpv4Fallback": 3,
        "Ipv6Only":            4,
}

```

Why not use those names?

Comment (solarkennedy) [reply executor/runtime/types/container.go:1338]: I don't know if we want those internal names to be exposed to the user, maybe.

Here is the doc with the spec I'm conforming to:
https://docs.google.com/document/d/1IJxY4LaiVYy5RqXDRtYSscvysae4bB7-3v4SSqvdNbE/edit

File: vpc/tool/assign3/assign_network.go [hunk 5]
```diff
@@ -32,10 +32,12 @@ type Arguments struct {
 	Jumbo              bool
 	Bandwidth          uint64
 	Burst              bool
+	NetworkMode        string
```
Comment (sargun) [inline vpc/tool/assign3/assign_network.go:35]: The titus executor and titus-vpc-tool are (effectively) atomically upgraded. So, you should get rid of the `AssignIPv6Address` logic here.

Comment (solarkennedy) [reply vpc/tool/assign3/assign_network.go:35]: I don't quite understand.
You want me to remove the `AssignIPv6Address` cli flag?

File: executor/runtime/types/container.go [hunk 6]
```diff
@@ -129,6 +129,7 @@ type TitusInfoContainer struct {
 	resources Resources
 
 	// VPC driver fields
+	networkMode        string
 	assignIPv6Address  bool
```
[the remark this answers is not in the document]
Comment (solarkennedy) [reply executor/runtime/types/container.go:133]: We can't get rid of it, this bool-as-attribute is still being set in many places.

I *could* "Fake" the networkMode setting:

```
if mode == Unset {
  if assignV6:
    mode = Ipv4AndV6
  }
}
```

But I don't really want to.
I would rather the networkMode stay untouched throughout

File: executor/runtime/types/container.go (hunk 4 above)
Comment (solarkennedy) [reply executor/runtime/types/container.go:1338]: Hmm. Like with another enum?

File: vpc/tool/assign3/assign_network.go (hunk 5 above)
Comment (sargun) [reply vpc/tool/assign3/assign_network.go:35]: Yeah, isn't `AssignIPv6Address` going to be subsumed by network mode?

File: executor/runtime/types/container.go (hunk 4 above)
Comment (sargun) [reply executor/runtime/types/container.go:1359]: or a map of nice-names to less-nice-names (and vice-versa)?

File: vpc/tool/assign3/assign_network.go (hunk 5 above)
Comment (solarkennedy) [reply vpc/tool/assign3/assign_network.go:35]: Eventually yes, but network mode will be Unknown (`0`) for a long time, we still need this bool to be in place to send in this data when network mode is 0.

File: executor/runtime/types/container.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (sargun) [reply executor/runtime/types/container.go:133]: I only see it being written to in boolPassthroughs. I'm suggesting that we replace the function here:
https://github.com/Netflix/titus-executor/blob/ba7a401b6e6820a8d1d030a8a5be0826e0cc34b5/executor/runtime/types/container.go#L524

and check if networkMode is set, and use that *first*, and then we add a TODO to remove this once we deprecate the old label.

[the remark this answers is not in the document]
Comment (solarkennedy) [reply executor/runtime/types/container.go:133]: Got it

**Comment by solarkennedy** [comment]
@sargun redone. `EffectiveNetworkMode()` is now the source of truth for what kind of network mode we want, and that is the only thing vpctool gets.

`EffectiveNetworkMode()` is computed based off of the input network mode and the attributes for backwards compat.

File: executor/runtime/docker/docker.go [hunk 7]
```diff
@@ -350,6 +351,25 @@ func maybeAddOptimisticDad(sysctl map[string]string) {
...
+	case titus.NetworkConfiguration_Ipv6AndIpv4Fallback.String():
+		// IPv6 with fallback means we only want v6 resolvers, which reduces
+		// the burden on TSA for ipv4 udp traffic
+		return []string{"fd00:ec2::253"}
+	case titus.NetworkConfiguration_Ipv6Only.String():
+		// True ipv6 only means we really can only have a v6
+		// resolver
+		return []string{"fd00:ec2::253"}
+	default:
+		// Any other situation means we can return the classic v4 resolver
```
Comment (sargun) [inline executor/runtime/docker/docker.go:368]: I'd // TODO this as maybe defaulting to V6 later, and treating V4 only as a specific case

File: executor/runtime/docker/docker.go [hunk 8]
```diff
@@ -987,7 +996,7 @@ func (r *DockerRuntime) Prepare(ctx context.Context) error { // nolint: gocyclo
 				"TITUS_SECCOMP_AGENT_HANDLE_PERF_SYSCALLS": "true",
 			})
 		}
-		if r.c.SeccompAgentEnabledForNetSyscalls() {
+		if r.c.SeccompAgentEnabledForNetSyscalls() || r.c.EffectiveNetworkMode() == titus.NetworkConfiguration_Ipv6AndIpv4Fallback.String() {
```
Comment (sargun) [inline executor/runtime/docker/docker.go:999]: Long term, should SeccompAgentEnabledForNetSyscalls go away?

File: executor/runtime/types/container.go [hunk 9]
```diff
@@ -219,6 +223,16 @@ func NewTitusInfoContainer(taskID string, titusInfo *titus.ContainerInfo, resour
 		c.extraUserContainers, c.extraPlatformContainers = NewExtraContainersFromPod(*c.pod)
 	}
 
+	pConfig := podCommon.Config{}
+	c.podConfig = &pConfig
```
Comment (sargun) [inline executor/runtime/types/container.go:227]: Any reason to break this into two separate lines?

File: executor/runtime/types/container.go [hunk 10]
```diff
@@ -810,6 +826,14 @@ func (c *TitusInfoContainer) MetatronCreds() *titus.ContainerInfo_MetatronCreds
 	return c.titusInfo.GetMetatronCreds()
 }
 
+func (c *TitusInfoContainer) EffectiveNetworkMode() string {
```
Comment (sargun) [inline executor/runtime/types/container.go:829]: I'd just add a comment, even though there is a more detailed comment on `computeEffectiveNetworkMode` of what this means.

File: executor/runtime/types/pod.go [hunk 11]
```diff
@@ -549,6 +557,9 @@ func (c *PodContainer) Runtime() string {
 	return DefaultOciRuntime
 }
 
+// SeccompAgentEnabledForNetSyscalls only represents whether a user set the SeccompAgentEnabledForNetSyscalls
```
Comment (sargun) [inline executor/runtime/types/pod.go:560]: If so, why would we leave this here and not always use network mode?

File: go.mod [hunk 12]
```diff
@@ -28,16 +28,19 @@ require (
 	github.com/docker/go-connections v0.4.0 // indirect
 	github.com/docker/go-units v0.4.0
 	github.com/ftrvxmtrx/fd v0.0.0-20150925145434-c6d800382fff
+	github.com/gogo/protobuf v1.3.2 // indirect
```
Comment (sargun) [inline go.mod:31]: What's pulling this in?

**Review by sargun — APPROVED** [review]

File: executor/runtime/docker/docker.go (hunk 8 above)
Comment (solarkennedy) [reply executor/runtime/docker/docker.go:999]: Yes, I just didn't want to break existing stuff.

Oh I see, no you are right, EffectiveNetworkMode() knows about that already.

File: executor/runtime/types/container.go (hunk 9 above)
Comment (solarkennedy) [reply executor/runtime/types/container.go:227]: nope

File: executor/runtime/types/pod.go (hunk 11 above)
Comment (solarkennedy) [reply executor/runtime/types/pod.go:560]: I gotta wait till we are not using the attribute everywhere?

I still need *some* function in here to tell me if the attribute is set.
Once spinnaker and stuff can set network mode, I'll depreciate usage of this attribute asap and remove these things.

File: go.mod (hunk 12 above)
Comment (solarkennedy) [reply go.mod:31]: My best guess is `github.com/Netflix/titus-kube-common` bump

We don't seem to be actually using it in the titus-executor codebase directly.
go mod tidy doesn't make it go away

Comment (solarkennedy) [reply go.mod:31]: Oh, k8s itself?
```kubernetes/go.mod:	github.com/gogo/protobuf v1.3.2```

**Pull request merged**